### DISCUSSION TITLE
Could not load edge function at `'/opt/build/repo/.netlify/edge-functions/server.js'`

### DISCUSSION THREAD

**Discussion by devraj** [netlify/remix-template#122] [Q&A] [answered]
I am using the `remix-template` to create a application (oh well a [template stack](https://github.com/anomaly/psychedelic-stack) which I can use across project), this is my second attempt at it (having finished a round of experimentation). Initially I created a project from this template and used 

```
netlify deploy --build --prod
```

as suggested in the `README` (without a `git` repository) and the deploy worked. 

I then recreated a project from a template, this time versioned in a `git` repository to establish a `git` based workflow with `netlify`, and everything worked as expected, the site deployed to `netlify` and I could see the `edge` functions logging on the console. Happy days, until I started making changes (primarily moving to `v2` compatibility for the remix project) and now the build won't find the edge functions.

<img width="564" alt="Screenshot 2023-06-27 at 11 36 40 am" src="https://github.com/netlify/remix-template/assets/5961/365c72fa-e889-41c9-b1f6-4308475bf7db">

Each builds fails with:

```
10:18:40 AM: Bundling of edge function failed                              
10:18:40 AM: ────────────────────────────────────────────────────────────────
10:18:40 AM: ​
10:18:40 AM:   Error message
10:18:40 AM:   Could not load edge function at '/opt/build/repo/.netlify/edge-functions/server.js'. More on the Edge Functions API at https://ntl.fyi/edge-api.
10:18:40 AM: ​
```

> I tried to update `remix` and all other outdated packages to see if that would make a difference but no luck there

Running `npm run start` which runs `netlify serve` results in it saying `Buffer` is not defined.

```
Packaging Edge Functions from .netlify/edge-functions directory:
 - server
ReferenceError: Buffer is not defined
    at node_modules/@remix-run/web-fetch/dist/lib.node.cjs (file:///Users/devraj/Work/Labs/psychedelic-stack/.netlify/edge-functions/server.js:85079:45)
    at file:///Users/devraj/Work/Labs/psychedelic-stack/.netlify/edge-functions/server.js:10:50
    at node_modules/@remix-run/node/dist/fetch.js (file:///Users/devraj/Work/Labs/psychedelic-stack/.netlify/edge-functions/server.js:86109:20)
```

Posts on `Stackoverflow` suggested that I install `buffer` which results in:

```
It appears you're using a module that is built in to node, but you installed it as a dependency which could cause problems. Please remove buffer, process before continuing.
```

Attempting to build using the `netlify` `cli` now fails with it not being able to load the edge functions

```
10:18:40 AM: ​
10:18:40 AM: Bundling of edge function failed                              
10:18:40 AM: ────────────────────────────────────────────────────────────────
10:18:40 AM: ​
10:18:40 AM:   Error message
10:18:40 AM:   Could not load edge function at '/opt/build/repo/.netlify/edge-functions/server.js'. More on the Edge Functions API at https://ntl.fyi/edge-api.
10:18:40 AM: ​
10:18:40 AM:   Error location
10:18:40 AM:   While bundling edge function
10:18:40 AM: ​
10:18:40 AM: Failed during stage 'building site': Build script returned non-zero exit code: 2 (https://ntl.fyi/exit-code-2)
10:18:40 AM:   Resolved config
```

What does work though is running `ntl dev` which works for development purposes but is not a solution for me to deploy the app to netlify via a `git` workflow.

I am not sure if this is something to do with the template or should I be asking this question on the main `netlify` forums. Any assistance or pointers are greatly appreciated.

**Comment by devraj** [comment]
Following posting this discussion, I decided to build the project ground up to see at what point the app deployment broke. It would turn out that it's the `loader` function in the `index` (or any of the pages) causes the failure of the edge functions being bundled.

My `loader` function is defined as:

```tsx
import type {
  V2_MetaFunction,
  LoaderFunction,
  LoaderArgs
} from "@remix-run/node";
import {
  json
} from "@remix-run/node";

export const loader: LoaderFunction = async ({ request }: LoaderArgs) => {
  return json({ message: "Hello World" });
}
```

I can also confirm that the `functions` template does not crash with loaders.

My conclusion is that it's something to do with the [Edge Function API](https://docs.netlify.com/edge-functions/api/#edge-function-types) and how `remix` defines functions. The missing piece must be the blue between `remix` functions and what edge functions expect?

I did find the [blog post outlining using edge functions with Remix](https://www.netlify.com/blog/how-to-use-remix-framework-with-edge-functions/) and I assume the template referenced in the blog post is the same that `init` function kicks off post creation of a `remix` project?

Wonder if it's worth providing a sample for developers to work from?

**Reply by nickytonline** [reply, marked as answer]
Since you're using edge functions, import from `@remix-run/deno`. Try that out and let me know if that gets it sorted.

**Reply by devraj** [reply]
Sure does, thank you for posting. 

For anyone else interested here's [my commit](https://github.com/anomaly/psychedelic-stack/commit/61a5969a6272bd78816a1e1549355c500872c46d) that fixed it for my deploy.
